spm.1
.Dd April 21, 2017
.Dt SPM 1
.Sh NAME
.Nm spm
.Nd simple password manager
.Sh SYNOPSIS
.Nm
.Cm add Oo Ar group/ Oc Ns Ar entry
.Nm
.Cm del Ar expression
.Nm
.Cm help
.Nm
.Cm list Oo Fl g Oc Op Ar group
.Nm
.Cm search Ar expression
.Nm
.Cm show Ar expression
.Sh DESCRIPTION
.Nm
is a fully POSIX shell compliant script using
.Xr gpg2 1
for encryption.
.sp
Entries are stored in a classical filesystem hierarchy where passwords
are encrypted as individual files and directories may be used as groups.
This way users can control access through file ownership, share entries
entries with symbolic links and more just like any other files.
.sp
Group and entry names or expressions are considered relative to the
storage root directoy.
.Ar expression
is interpreted as basic regular expression according to
.Xr re_format 7 .
The following commands are available:
.Bl -tag -width Ds
.It Cm add
Prompt for a password if standard output is a terminal otherwise read
from standard input and store it as
.Ar entry .
If
.Ar group
is specified, non-existing groups are created automatically and the
password is stored under
.Ar group/ Ns Op Ar ../ .
.It Cm del
Ask for confirmation and delete the entry matching
.Ar expression
on success.
.It Cm help
Print a short usage and refer to this manual page.
.It Cm list
Print a recursive, indented tree-style list of all entries in
.Ar group .
If
.Ar group
is omitted, the entire storage root is listed.
If the
.Fl g
flag is passed only subgroups are shown.
.It Cm search
Print the absoloute paths of all entries matching
.Ar expression
one per line.
.It Cm show
Print the password to standard output if
.Ar expression
uniquely matches an entry and prompt for the decryption passphrase if
it is not cached by
.Xr gpg-agent 1
already.
.El
.Sh ENVIRONMENT
The following environment variables will be used by
.Nm
if they are set and not empty:
.Bl -tag -width Ds
.It Ev PASSWORD_STORE_DIR
The storage root directory.
This should be an absoloute path.
.It Ev PASSWORD_STORE_KEY
Key ID used for de/encryption.
Refer to
.Xr gpg2 1
for accepted formats.
.El
.Sh FILES
.Bl -tag -width Ds
.It Pa ~/.spm/
The default storage root directory.
.El
.Sh EXIT STATUS
.Nm
exits 1 if an error occurs otherwise 0.
.Sh EXAMPlES
Add Alice's password to the group
.Em users :
.Bd -literal -offset indent
$ spm add users/alice
Password for 'users/alice':
.Ed
.sp
Make sure she is an admin as well:
.Bd -literal -offset indent
$ (cd ~/.spm
> mkdir admins
> ln -s ../members/alice admins/alice)
.Ed
.sp
Copy her password into the X selection ready to be pasted somewhere:
.Bd -literal -offset indent
$ spm show admins/alice | xclip -i
.Ed
.sp
Find all admins starting with
.Dq a
or
.Dq b :
.Bd -literal -offset indent
$ spm search admins/[ab]
/home/alice/.spm/admins/alice
/home/alice/.spm/admins/bob
.Ed
.sp
List all users:
.Bd -literal -offset indent
$ spm list users
/home/alice/.spm/users
|-- alice
|-- bob
|-- temporary/
|   |-- tim
|   `-- tom
.Ed
.sp
Find entries interactively using
.Xr dmenu 1 :
.Bd -literal -offset indent
$ spm show "$(spm search . | dmenu -l 10)"
.Ed
.Sh SEE ALSO
.Xr dmenu 1,
.Xr gpg2 1 ,
.Xr gpg-agent 1 ,
.Xr tpm 1 ,
.Xr tree 1 ,
.Xr re_format 7
.Sh CAVEATS
Groups have to be deleted manually using
.Xr rmdir 1 .
.Sh AUTHORS
.Nm
is written by
.An Klemens Nanni Aq Mt kl3@posteo.org .


spm.sh
#!/bin/sh
# Copyright (C) 2013-2016 Sören Tempel
# Copyright (C) 2016, 2017 Klemens Nanni <kl3@posteo.org>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

set -eu
umask 077

## Variables
GPG_OPTS='--quiet --yes --batch'
STORE_DIR="${PASSWORD_STORE_DIR:-${HOME}/.spm}"

## Helper
usage() {
	cat 1>&2 <<-EOF
	${1:+Error: ${1}}
	USAGE: ${0##*/} add|del|list [-g]|search|show|help [[group/]name|expression]
	See spm(1) for more information.
	EOF

	exit ${1:+1}
}

check() {
	[ $(printf '%s' "${entry}" | wc -l) -eq 0 ] ||
		usage 'ambigious expression'

	[ -n "${entry}" ] || usage 'no such entry'
}

gpg() {
	if [ -z "${PASSWORD_STORE_KEY}" ]; then
		gpg2 ${GPG_OPTS} --default-recipient-self "${@}"
	else
		gpg2 ${GPG_OPTS} --recipient "${PASSWORD_STORE_KEY}" "${@}"
	fi
}

readpw() {
	[ -t 0 ] && stty -echo && printf '%s' "${1}"
	IFS= read -r "${2}"
	[ -z "${2}" ] && usage 'empty password'
}

find() {
	command find "${STORE_DIR}" -type f -o -type l | grep -Gie "${1}"
}

view() {
	less -EiKRX
}

## Commands
add() {
	[ -e "${STORE_DIR}"/"${1}" ] && usage 'entry already exists'

	password=
	readpw "Password for '${1}': " password
	[ -t 0 ] && printf '\n'

	group="${1%/*}"
	[ "${group}" = "${1}" ] && group=

	mkdir -p "${STORE_DIR}"/"${group}" &&
		printf '%s\n' "${password}" |
			gpg --encrypt --output "${STORE_DIR}"/"${1}"
}

list() {
	[ -d "${STORE_DIR}"/"${1:-}" ] || usage 'no such group'

	tree ${gflag:+-d} -Fx -- "${STORE_DIR}"/"${1:-}" | view
}

del() {
	entry=$(find "${1}" | head -n2)
	check; rm -i "${entry}" && printf '\n'
}

search() {
	find "${1}" | view
}

show() {
	entry=$(find "${1}" | head -n2)
	check; gpg --decrypt "${entry}"
}

## Parse input
[ ${#} -eq 0 ] || [ ${#} -gt 3 ] ||
[ ${#} -eq 3 ] && [ "${1:-}" != list ] &&
	usage 'wrong number of arguments'

case "${1}" in
	add|del|search|show)
		[ -z "${2:-}" ] && usage 'empty name'
		${1} "${2}"
		;;
	list)
		if [ "${2:-}" = -g ] && [ ${#} -le 3 ]; then
			gflag=1; shift 1
		elif [ ${#} -gt 2 ]; then
			usage 'too many arguments'
		fi
		list "${2:-}"
		;;
	help)
		usage
		;;
	*)
		usage 'invalid command'
		;;
esac
